File: src/mcp/testdata/echo_server.ts
/**
 * Test fixture: a tiny MCP server over stdio (newline-delimited JSON-RPC).
 *
 * Speaks just enough protocol for `client.test.ts` — initialize, tools/list in TWO
 * pages (so the cursor loop is exercised), tools/call — and deliberately misbehaves
 * on demand, because the client's whole contract is what happens when a server
 * does not cooperate.
 *
 * Run with `bun`: it reads stdin and writes stdout/stderr, nothing else.
 *
 * Tools:
 *   echo   {text}  — readOnlyHint; returns the text plus structuredContent
 *   scream {text}  — annotated as a write; uppercases
 *   boom   {}      — returns an isError RESULT (a tool failure is data)
 *   die    {}      — writes to stderr and exits the process MID-CALL, never replying
 *   slow   {}      — never replies at all, and stays alive doing it
 *   loose  {q}     — inputSchema missing `type: "object"`, so only the lenient path keeps it
 *
 * Flags:
 *   --deaf   read stdin forever, answer nothing (a server that starts and hangs)
 *   --noise  print a non-JSON banner to stdout before the handshake
 */

// This fixture imports nothing, and top-level `await` is only legal in a module —
// so say so explicitly. Deno called every file a module; tsc does not.
export {};

const args = process.argv.slice(2);
const DEAF = args.includes("--deaf");
const NOISE = args.includes("--noise");

function reply(message: unknown): void {
  console.log(JSON.stringify(message));
}

const PAGE1 = [
  {
    name: "echo",
    description: "Echo the text back.\nSecond line that the prompt section must drop.",
    inputSchema: {
      type: "object",
      properties: { text: { type: "string" } },
      required: ["text"],
    },
    annotations: { readOnlyHint: true },
  },
];

const PAGE2 = [
  {
    name: "scream",
    description: "Echo the text back, LOUDLY.",
    inputSchema: { type: "object", properties: { text: { type: "string" } }, required: ["text"] },
    annotations: { readOnlyHint: false, destructiveHint: true },
  },
  { name: "boom", description: "Always fails.", inputSchema: { type: "object", properties: {} } },
  {
    name: "die",
    description: "Kills the server.",
    inputSchema: { type: "object", properties: {} },
  },
  { name: "slow", description: "Never answers.", inputSchema: { type: "object", properties: {} } },
  // Missing `type: "object"` — the SDK's ToolSchema rejects it; the client's
  // lenient fallback must still list it, because the tool is callable.
  {
    name: "loose",
    description: "Advertised with a sloppy schema.",
    inputSchema: { properties: { q: { type: "string" } } },
  },
  // No name at all: not callable even in principle, so the client drops it.
  { description: "an entry with no name" },
];

function callTool(id: number | undefined, name: string, args: Record<string, unknown>): void {
  if (name === "die") {
    // Mid-call death: no reply, a diagnostic on stderr, gone.
    console.error("echo-fixture: asked to die, taking the server down");
    process.exit(3);
  }
  if (name === "slow") return; // alive, and never answering

  const result = name === "echo"
    ? {
      content: [{ type: "text", text: String(args.text) }],
      structuredContent: { echoed: args.text },
    }
    : name === "scream"
    ? { content: [{ type: "text", text: String(args.text).toUpperCase() }] }
    : name === "boom"
    ? { content: [{ type: "text", text: "kaboom" }], isError: true }
    : name === "loose"
    ? { content: [{ type: "text", text: `q=${String(args.q)}` }] }
    : { content: [{ type: "text", text: `no such tool: ${name}` }], isError: true };
  reply({ jsonrpc: "2.0", id, result });
}

function handle(message: {
  id?: number;
  method?: string;
  params?: { cursor?: string; name?: string; arguments?: Record<string, unknown> };
}): void {
  if (DEAF) return;
  if (message.method === "initialize") {
    reply({
      jsonrpc: "2.0",
      id: message.id,
      result: {
        protocolVersion: "2025-06-18",
        capabilities: { tools: {} },
        serverInfo: { name: "echo-fixture", version: "0" },
      },
    });
  } else if (message.method === "tools/list") {
    const second = message.params?.cursor === "page2";
    reply({
      jsonrpc: "2.0",
      id: message.id,
      result: second ? { tools: PAGE2 } : { tools: PAGE1, nextCursor: "page2" },
    });
  } else if (message.method === "tools/call") {
    callTool(message.id, message.params?.name ?? "", message.params?.arguments ?? {});
  } else if (message.id !== undefined) {
    reply({ jsonrpc: "2.0", id: message.id, error: { code: -32601, message: "method not found" } });
  }
  // notifications (initialized) need no reply
}

if (NOISE) console.log("echo-fixture starting up (this line is not JSON)");

let buffer = "";
for await (const chunk of Bun.stdin.stream().pipeThrough(new TextDecoderStream())) {
  buffer += chunk;
  for (;;) {
    const newline = buffer.indexOf("\n");
    if (newline < 0) break;
    const line = buffer.slice(0, newline).trim();
    buffer = buffer.slice(newline + 1);
    if (line) handle(JSON.parse(line));
  }
}

File: scripts/fixtures/mcp-process/echo-server.py
#!/usr/bin/env python3
"""A minimal resident MCP server over stdio, for the `mcp-subprocess` tests.

Speaks line-delimited JSON-RPC 2.0: `initialize`, `tools/list`, `tools/call`. It never reaches the
network and never writes outside `MCP_FIXTURE_RECORD`.

Environment knobs, so one fixture covers every case the host must survive:
  MCP_FIXTURE_RECORD   append one line per process start (the spawn counter a test reads)
  MCP_FIXTURE_DIE_MS   exit(1) this many milliseconds after start — the crash-loop fixture
  MCP_FIXTURE_ACTIONS  emit one `bough/actions` notification right after `initialize`
  MCP_FIXTURE_TOOL     the name of the one tool listed (default `echo`)
"""

import json
import os
import sys
import threading
import time


def send(msg):
    sys.stdout.write(json.dumps(msg) + "\n")
    sys.stdout.flush()


def main():
    record = os.environ.get("MCP_FIXTURE_RECORD")
    if record:
        with open(record, "a") as f:
            f.write("start %d\n" % os.getpid())

    die_ms = os.environ.get("MCP_FIXTURE_DIE_MS")
    if die_ms:
        def die():
            time.sleep(int(die_ms) / 1000.0)
            os._exit(1)
        threading.Thread(target=die, daemon=True).start()

    tool = os.environ.get("MCP_FIXTURE_TOOL", "echo")

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            req = json.loads(line)
        except ValueError:
            continue
        method, rid = req.get("method"), req.get("id")
        if method == "initialize":
            send({"jsonrpc": "2.0", "id": rid, "result": {
                "protocolVersion": "2025-06-18",
                "capabilities": {"tools": {}},
                "serverInfo": {"name": "echo-fixture", "version": "0.1.0"},
            }})
            if os.environ.get("MCP_FIXTURE_ACTIONS"):
                send({"jsonrpc": "2.0", "method": "bough/actions", "params": {"actions": [
                    {"kind": "hint", "agent": "sol", "text": "a resident process said so"}
                ]}})
        elif method == "tools/list":
            send({"jsonrpc": "2.0", "id": rid, "result": {"tools": [{
                "name": tool,
                "description": "echoes its argument back",
                "inputSchema": {"type": "object", "properties": {"text": {"type": "string"}}},
            }]}})
        elif method == "tools/call":
            args = (req.get("params") or {}).get("arguments") or {}
            send({"jsonrpc": "2.0", "id": rid, "result": {
                "content": [{"type": "text", "text": args.get("text", "")}],
                "isError": False,
            }})
        elif rid is not None:
            send({"jsonrpc": "2.0", "id": rid,
                  "error": {"code": -32601, "message": "no such method: %s" % method}})


if __name__ == "__main__":
    main()
